cmplib/cmpcl_read.c

```diff
@@ -971,15 +993,61 @@ static int cmp_pkibody_errmsgcnt_parse_der( cmp_ErrorMsgContent *emc,
     if ((ret = cmp_pkibody_PKIStatusInfo_parse_der( &emc->pKIStatusInfo, p, p+len)))
         return(ret); /* TODO improve */
 
+    p += len;
+    if(p == end)
+    	return 0;
+
+    /*errorCode              INTEGER           OPTIONAL,*/
+    unsigned char *tmpp = p;
 
-    if( p+len != end )
+    if( ( ret = mbedtls_asn1_get_tag( &tmpp, end, &len,	MBEDTLS_ASN1_INTEGER ) ) == 0 )
     {
-        if( p+len < end ) {
-            CMPDBGS("Additional errorCode and/or errorDetails in errorMsgBody ignored");
-        } else {
-            return( MBEDTLS_ERR_X509_INVALID_FORMAT + MBEDTLS_ERR_ASN1_LENGTH_MISMATCH );
-        }
+		if( ( ret = mbedtls_asn1_get_int( &p, end, &emc->errorCode ) ) != 0 )
+		{
+			MBEDTLS_ERR( ret, err_buf, CMPCL_ER_BUF_LEN );
+			return( CMPCL_ERR_ASN1_PARSING );
+		}
+		CMPERRV( "Response with Error Code: %d", emc->errorCode );
+		if(p == end)
+			return 0;
     }
+
+    /*
+            errorDetails           PKIFreeText       OPTIONAL
+               PKIFreeText ::= SEQUENCE SIZE (1..MAX) OF UTF8String
+      */
+	tmpp = p;
+	if( ( ret = mbedtls_asn1_get_tag( &tmpp, end, &len,
+				   MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE ) ) == 0 )
+	{
+	   len += tmpp-p;
+	   emc->errorDetails = (mbedtls_asn1_sequence *) mbedtls_calloc(1,
+												sizeof(mbedtls_asn1_sequence));
+	   if ( ( ret = mbedtls_asn1_get_sequence_of( &p, p+len,
+			   	   	   emc->errorDetails,
+					   MBEDTLS_ASN1_UTF8_STRING)))
+	   {
+		MBEDTLS_ERR( ret, err_buf, CMPCL_ER_BUF_LEN );
+		return( CMPCL_ERR_ASN1_PARSING );
+
+	   }
+	   mbedtls_asn1_sequence *ErrorDetails = emc->errorDetails;
+	   while(ErrorDetails)
+	   {
+		CMPERRV("Response Error Details: %.*s",
+			   (int) ErrorDetails->buf.len,
+			   ErrorDetails->buf.p);
+		ErrorDetails = ErrorDetails->next;
+	   }
+	}
```

Comment:
To avoid redundancy, better factor out this code section and essentially the same code in `cmp_pkibody_PKIStatusInfo_parse_der` into an auxiliary function that handles parsing and printing of `PKIFreeText` in a generic way.